core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -364,24 +376,48 @@ private void finishAsFailed(Exception exception) {
+        void performOn(ShardRouting replica, R replicaRequest, ActionListener<ReplicaResponse> listener);
 
         /**
-         * Fail the specified shard, removing it from the current set of active shards
+         * Fail the specified shard if needed, removing it from the current set
+         * of active shards. Whether a failure is needed is left up to the
+         * implementation.
+         *
          * @param replica          shard to fail
          * @param primaryTerm      the primary term of the primary shard when requesting the failure
          * @param message          a (short) description of the reason
          * @param exception        the original exception which caused the ReplicationOperation to request the shard to be failed
          * @param onSuccess        a callback to call when the shard has been successfully removed from the active set.
          * @param onPrimaryDemoted a callback to call when the shard can not be failed because the current primary has been demoted
-*                         by the master.
+         *                         by the master.
          * @param onIgnoredFailure a callback to call when failing a shard has failed, but it that failure can be safely ignored and the
+         *
+         * @return true if the shard was failed, false otherwise
+         */
+        boolean failShardIfNeeded(ShardRouting replica, long primaryTerm, String message, Exception exception, Runnable onSuccess,
+                                  Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure);
+
+        /**
+         * Marks shard copy as stale, removing its allocation id from the set
+         * of in-sync allocation ids. Will only mark the shard copy stale when
+         * absolutely necessary. For example, failed write operations should
+         * mark the shard as stale, but global checkpoint syncs that fail
+         * should not.
+         *
+         * @param shardId          shard id
+         * @param allocationId     allocation id to remove from the set of in-sync allocation ids
+         * @param primaryTerm      the primary term of the primary shard when requesting the failure
+         * @param onSuccess        a callback to call when the allocation id has been successfully removed from the in-sync set.
+         * @param onPrimaryDemoted a callback to call when the request failed because the current primary was already demoted
+         *                         by the master.
+         * @param onIgnoredFailure a callback to call when the request failed, but the failure can be safely ignored.
          */
-        void failShard(ShardRouting replica, long primaryTerm, String message, Exception exception, Runnable onSuccess,
-                       Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure);
+        void markShardCopyAsStaleIfNeeded(ShardId shardId, String allocationId, long primaryTerm, Runnable onSuccess,
```

Comment:
We need two because there are situations when a replica shard needs to be marked as stale without a write action, I originally had this and then the tests that restart ES exposed that they still need to be able to mark it as stale without any actions.